### PR TITLE
Fix/ota 831/retry download

### PR DESCRIPTION
**Pull request by patriotyk**

### CODE REVIEW

File: src/libaktualizr/uptane/uptane_network_test.cc [hunk 1]
```diff
@@ -113,6 +115,30 @@ TEST(UptaneNetwork, no_errors_sqlite) {
   EXPECT_TRUE(doTestInit(StorageType::kSqlite, "noerrors", "noerrors"));
 }
 
+TEST(UptaneNetwork, DownloadFailure) {
+  TemporaryDirectory temp_dir;
+  conf.storage.path = temp_dir.Path();
+  conf.provision.expiry_days = "noerrors";
+  conf.provision.primary_ecu_serial = "noerrors";
+  conf.provision.primary_ecu_hardware_id = "download_failure";
```
Comment (patrickvacek) [inline src/libaktualizr/uptane/uptane_network_test.cc]: Is there a significance to "download_failure" here? I don't see it used elsewhere.

File: tests/fake_http_server/fake_uptane_server.py [hunk 2]
```diff
@@ -8,7 +8,24 @@
 
 class Handler(BaseHTTPRequestHandler):
     def do_GET(self):
-        pass
+        if self.path.startswith("/repo/targets/noerrors"):
```
Comment (patrickvacek) [inline tests/fake_http_server/fake_uptane_server.py]: Should this really be for "noerrors"? Won't this get triggered for all the network tests?

**Comment by patrickvacek** [comment]
I merged my large PR, so this needs to be rebased. I fixed one of the flaky tests, though, so at least that's one less thing to worry about.

**Comment by patriotyk** [comment]
I can't look in to the Jenkins

**Comment by patrickvacek** [comment]
> I can't look in to the Jenkins

Sorry. It's hsm_prov again. I don't understand why it is failing, but here's the log:

```
114/116 Test #102: hsm_prov_test .......................................***Failed   64.63 sec
> Running /home/laurent/jenkins-slave/jenkins-home/workspace/aktualizr_PR-1001-7RE67M5RGS372BWYW464TROUSU44ZFUJ23LZBJSSS6WN527WU2YA@3/build-coverage/src/aktualizr_primary/aktualizr --config /tmp/tmpgh_d1_b3/conf.d --loglevel 0 --running-mode once
> Running /home/laurent/jenkins-slave/jenkins-home/workspace/aktualizr_PR-1001-7RE67M5RGS372BWYW464TROUSU44ZFUJ23LZBJSSS6WN527WU2YA@3/build-coverage/src/aktualizr_info/aktualizr-info --config /tmp/tmpgh_d1_b3/conf.d
> Running /home/laurent/jenkins-slave/jenkins-home/workspace/aktualizr_PR-1001-7RE67M5RGS372BWYW464TROUSU44ZFUJ23LZBJSSS6WN527WU2YA@3/build-coverage/src/aktualizr_info/aktualizr-info --config /tmp/tmpgh_d1_b3/conf.d
Device has not yet provisioned (as expected). Running aktualizr-cert-provider.
> Running /home/laurent/jenkins-slave/jenkins-home/workspace/aktualizr_PR-1001-7RE67M5RGS372BWYW464TROUSU44ZFUJ23LZBJSSS6WN527WU2YA@3/build-coverage/src/cert_provider/aktualizr-cert-provider -c **** -l / -r -s -u -g /tmp/tmpgh_d1_b3/conf.d/20-sota_hsm_prov.toml
> Running /home/laurent/jenkins-slave/jenkins-home/workspace/aktualizr_PR-1001-7RE67M5RGS372BWYW464TROUSU44ZFUJ23LZBJSSS6WN527WU2YA@3/scripts/export_to_hsm.sh
> Running pkcs11-tool --module=/usr/lib/softhsm/libsofthsm2.so -O --login --pin 1234
> Running softhsm2-util --show-slots
> Running /home/laurent/jenkins-slave/jenkins-home/workspace/aktualizr_PR-1001-7RE67M5RGS372BWYW464TROUSU44ZFUJ23LZBJSSS6WN527WU2YA@3/build-coverage/src/aktualizr_primary/aktualizr --config /tmp/tmpgh_d1_b3/conf.d --loglevel 0 --running-mode once
> Running /home/laurent/jenkins-slave/jenkins-home/workspace/aktualizr_PR-1001-7RE67M5RGS372BWYW464TROUSU44ZFUJ23LZBJSSS6WN527WU2YA@3/build-coverage/src/aktualizr_info/aktualizr-info --config /tmp/tmpgh_d1_b3/conf.d
> Running /home/laurent/jenkins-slave/jenkins-home/workspace/aktualizr_PR-1001-7RE67M5RGS372BWYW464TROUSU44ZFUJ23LZBJSSS6WN527WU2YA@3/build-coverage/src/aktualizr_info/aktualizr-info --config /tmp/tmpgh_d1_b3/conf.d
> Running /home/laurent/jenkins-slave/jenkins-home/workspace/aktualizr_PR-1001-7RE67M5RGS372BWYW464TROUSU44ZFUJ23LZBJSSS6WN527WU2YA@3/build-coverage/src/aktualizr_info/aktualizr-info --config /tmp/tmpgh_d1_b3/conf.d
> Running /home/laurent/jenkins-slave/jenkins-home/workspace/aktualizr_PR-1001-7RE67M5RGS372BWYW464TROUSU44ZFUJ23LZBJSSS6WN527WU2YA@3/build-coverage/src/aktualizr_info/aktualizr-info --config /tmp/tmpgh_d1_b3/conf.d
> Running /home/laurent/jenkins-slave/jenkins-home/workspace/aktualizr_PR-1001-7RE67M5RGS372BWYW464TROUSU44ZFUJ23LZBJSSS6WN527WU2YA@3/build-coverage/src/aktualizr_info/aktualizr-info --config /tmp/tmpgh_d1_b3/conf.d
> Running /home/laurent/jenkins-slave/jenkins-home/workspace/aktualizr_PR-1001-7RE67M5RGS372BWYW464TROUSU44ZFUJ23LZBJSSS6WN527WU2YA@3/build-coverage/src/aktualizr_info/aktualizr-info --config /tmp/tmpgh_d1_b3/conf.d
> Running /home/laurent/jenkins-slave/jenkins-home/workspace/aktualizr_PR-1001-7RE67M5RGS372BWYW464TROUSU44ZFUJ23LZBJSSS6WN527WU2YA@3/build-coverage/src/aktualizr_info/aktualizr-info --config /tmp/tmpgh_d1_b3/conf.d
> Running /home/laurent/jenkins-slave/jenkins-home/workspace/aktualizr_PR-1001-7RE67M5RGS372BWYW464TROUSU44ZFUJ23LZBJSSS6WN527WU2YA@3/build-coverage/src/aktualizr_info/aktualizr-info --config /tmp/tmpgh_d1_b3/conf.d
Provisioning failed: Reading config: "/tmp/tmpgh_d1_b3/conf.d/20-sota_hsm_prov.toml"
Use existing SQL storage: "/tmp/tmpgh_d1_b3/sql.db"
Storage backend: Sqlite
Device ID: ornate-kuerbiskernbrot-619
Primary ecu serial ID: 1dd24927a4a01d29edea6eb731bb22aef856b5df036df7211cd027738ae4e018
Primary ecu hardware ID: 074b1a25f796
Provisioned on server: no
Fetched metadata: no
```

I restarted it in case it is just transient, but since it's happened two or three times now, it probably isn't.

**Comment by patrickvacek** [comment]
I've narrowed the problem down to the last commit about fixing valgrind complaints. Before that change, hsm provisioning works. I'm still looking into it.

It may also need a fix for reboot detection. (FYI @lbonn) At least when I test locally, I have to add something like this to my config:
```
[bootloader]
reboot_sentinel_dir = "/home/patrick/Code/aktualizr/tests/hsm_test/"
```
Otherwise it complains with `boost::filesystem::status: Permission denied: "/var/run/aktualizr-session/need_reboot"`.

**Comment by lbonn** [comment]
It looks like it is the same curl bug that @OYTIS initially bumped into (see TODO in `httpclient.cc`).

It's actually not an unknown bug anymore, I tracked it down to https://github.com/curl/curl/issues/2829.

For current distributions, the workaround is still required but can probably be shortened to:

```
if (pkcs11_key) {
    curlEasySetoptWrapper(curl_put, CURLOPT_SSLENGINE, "pkcs11");
}
```

This line should be enough in all cases and I suggest we'd put that in an internal wrapper for cloning curl handles to refactor it.

**Comment by patrickvacek** [comment]
> It looks like it is the same curl bug that @OYTIS initially bumped into (see TODO in `httpclient.cc`).

Agreed. I'll update this and try to take care of it since i'm already in the middle of it.

File: src/libaktualizr/http/httpclient.cc [hunk 3]
```diff
@@ -168,27 +168,22 @@ void HttpClient::setCerts(const std::string& ca, CryptoSource ca_source, const s
 }
 
 HttpResponse HttpClient::post(const std::string& url, const Json::Value& data) {
-  curlEasySetoptWrapper(curl, CURLOPT_URL, url.c_str());
-  curlEasySetoptWrapper(curl, CURLOPT_POST, 1);
+  CURL* curl_post = Utils::curlDupHandleWrapper(curl, pkcs11_key);
+  curlEasySetoptWrapper(curl_post, CURLOPT_URL, url.c_str());
+  curlEasySetoptWrapper(curl_post, CURLOPT_POST, 1);
+  if (pkcs11_key) {
```
Comment (lbonn) [inline src/libaktualizr/http/httpclient.cc]: Should be unneeded here (in Utils::curlDupHandleWrapper)

Comment (patrickvacek) [reply src/libaktualizr/http/httpclient.cc]: Yep, that was a mistake. Fixed.

**Comment by patriotyk** [comment]
Looks good for me

**Pull request merged**